File: AlerteTCL/Services/ParcRelaisService.swift
import Foundation
import CoreLocation

// MARK: - Decoding types (GeoJSON SYTRAL P+R)

private struct ParcRelaisStaticResponse: Codable {
    let features: [ParcRelaisStaticFeature]
}
private struct ParcRelaisStaticFeature: Codable {
    let geometry: ParcRelaisGeometry
    let properties: ParcRelaisStaticProperties
}
private struct ParcRelaisRealtimeResponse: Codable {
    let features: [ParcRelaisRealtimeFeature]
}
private struct ParcRelaisRealtimeFeature: Codable {
    let properties: ParcRelaisRealtimeProperties
}
private struct ParcRelaisGeometry: Codable {
    let coordinates: [[Double]]   // MultiPoint → premier point
    var coordinate: CLLocationCoordinate2D? {
        guard let first = coordinates.first, first.count >= 2 else { return nil }
        return CLLocationCoordinate2D(latitude: first[1], longitude: first[0])
    }
}
private struct ParcRelaisStaticProperties: Codable {
    let id: String
    let nom: String
    let capacite: Int?
    let placeHandi: Int?
    let horaires: String?
    let pSurv: Bool?
    enum CodingKeys: String, CodingKey {
        case id, nom, capacite, horaires
        case placeHandi = "place_handi"
        case pSurv = "p_surv"
    }
}
private struct ParcRelaisRealtimeProperties: Codable {
    let id: String
    let nbTotPlaceDispo: Int?
    enum CodingKeys: String, CodingKey {
        case id
        case nbTotPlaceDispo = "nb_tot_place_dispo"
    }
}

// MARK: - Service

actor ParcRelaisService {
    static let shared = ParcRelaisService()
    private init() {}

    private let staticURL   = NetworkConfiguration.proxyBaseURL + "/parc-relais?f=application/json&limit=100&sortby=gid"
    private let realtimeURL = NetworkConfiguration.proxyBaseURL + "/parc-relais-tr?f=application/json&limit=100&sortby=gid"

    // Cache statique — 24h
    private var cachedStatic: [Parking]?
    private var cacheTimestamp: Date?
    private let cacheValidity: TimeInterval = 86_400

    // Cache temps réel — 60s
    private var lastRealtimeFetch: Date?
    private let realtimeValidity: TimeInterval = 60

    // MARK: - Public API

    /// Retourne les P+R fusionnés avec les données temps réel disponibles.
    func fetchParcRelais(forceRefresh: Bool = false) async throws -> [Parking] {
        let now = Date()

        // 1. Données statiques (cache 24h)
        let staticList: [Parking]
        if !forceRefresh,
           let cached = cachedStatic,
           let ts = cacheTimestamp,
           now.timeIntervalSince(ts) < cacheValidity {
            staticList = cached
        } else {
            staticList = try await fetchStatic()
            cachedStatic = staticList
            cacheTimestamp = now
        }

        // 2. Données temps réel (cache 60s)
        let shouldRefreshRT = forceRefresh
            || lastRealtimeFetch == nil
            || now.timeIntervalSince(lastRealtimeFetch!) >= realtimeValidity

        guard shouldRefreshRT else { return staticList }

        let rtMap = (try? await fetchRealtimeMap()) ?? [:]
        lastRealtimeFetch = now

        // 3. Fusionner — reconstruire Parking avec dispo si disponible
        return staticList.map { pr in
            // Extraire l'id brut (on a préfixé "parc-relais-" dans l'init)
            let rawId = String(pr.id.dropFirst("parc-relais-".count))
            guard let dispo = rtMap[rawId] else { return pr }
            return Parking(
                parcRelaisId: rawId,
                nom: pr.nom,
                coordinate: pr.coordinate,
                capacite: pr.capaciteTotale,
                placesHandicap: pr.nbPmr ?? 0,
                horaires: pr.horaires,
                surveille: pr.surveille ?? false,
                placesDisponibles: dispo
            )
        }
    }

    // MARK: - Private helpers

    private func fetchStatic() async throws -> [Parking] {
        guard let url = URL(string: staticURL) else { throw ParcRelaisError.invalidURL }
        let (data, _) = try await NetworkConfiguration.shared.data(for: NetworkConfiguration.request(url: url, timeout: NetworkConfiguration.sharedTimeout))
        let resp = try JSONDecoder().decode(ParcRelaisStaticResponse.self, from: data)
        return resp.features.compactMap { feature -> Parking? in
            guard let coord = feature.geometry.coordinate else { return nil }
            let p = feature.properties
            return Parking(
                parcRelaisId: p.id,
                nom: p.nom,
                coordinate: coord,
                capacite: p.capacite ?? 0,
                placesHandicap: p.placeHandi ?? 0,
                horaires: p.horaires,
                surveille: p.pSurv ?? false,
                placesDisponibles: nil
            )
        }
    }

    private func fetchRealtimeMap() async throws -> [String: Int] {
        guard let url = URL(string: realtimeURL) else { throw ParcRelaisError.invalidURL }
        let (data, _) = try await NetworkConfiguration.shared.data(for: NetworkConfiguration.request(url: url, timeout: NetworkConfiguration.sharedTimeout))
        let resp = try JSONDecoder().decode(ParcRelaisRealtimeResponse.self, from: data)
        var map: [String: Int] = [:]
        for feature in resp.features {
            if let dispo = feature.properties.nbTotPlaceDispo {
                map[feature.properties.id] = dispo
            }
        }
        return map
    }
}

enum ParcRelaisError: LocalizedError {
    case invalidURL
    var errorDescription: String? { "URL invalide" }
}

File: shared/src/commonMain/kotlin/com/alertetcl/shared/services/ParcRelaisService.kt
package com.alertetcl.shared.services

import com.alertetcl.shared.geo.LatLng
import com.alertetcl.shared.models.Parking
import com.alertetcl.shared.models.ParkingState
import com.alertetcl.shared.models.ParkingType
import com.alertetcl.shared.network.ApiError
import com.alertetcl.shared.network.HttpClientProvider
import com.alertetcl.shared.network.NetworkConfiguration
import com.alertetcl.shared.network.dto.ParcRelaisRealtimeResponse
import com.alertetcl.shared.network.dto.ParcRelaisStaticResponse
import com.alertetcl.shared.network.safeDecode
import com.alertetcl.shared.network.safeRequest
import com.alertetcl.shared.util.AppLogger
import io.ktor.client.call.body
import io.ktor.client.plugins.timeout
import io.ktor.client.request.get
import io.ktor.http.HttpStatusCode
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.datetime.Clock

class ParcRelaisService {

    private val client = HttpClientProvider.client
    private val staticUrl   = NetworkConfiguration.PROXY_BASE_URL + "/parc-relais?f=application/json&limit=100&sortby=gid"
    private val realtimeUrl = NetworkConfiguration.PROXY_BASE_URL + "/parc-relais-tr?f=application/json&limit=100&sortby=gid"

    private val mutex = Mutex()
    private var cachedStatic: List<Parking>? = null
    private var staticTimestamp = 0L
    private val staticValidity = 86_400L

    private var lastRtTimestamp = 0L
    private val rtValidity = 60L
    private var lastMerged: List<Parking>? = null

    suspend fun fetchParcRelais(forceRefresh: Boolean = false): List<Parking> {
        val now = Clock.System.now().epochSeconds

        val staticList = mutex.withLock {
            val cached = cachedStatic
            if (!forceRefresh && cached != null && now - staticTimestamp < staticValidity) cached else null
        } ?: run {
            val list = fetchStatic()
            mutex.withLock { cachedStatic = list; staticTimestamp = now }
            list
        }

        val needRt = forceRefresh || (now - lastRtTimestamp >= rtValidity)
        if (!needRt) return lastMerged ?: staticList

        // Un échec temps réel ne doit pas écraser les disponibilités connues par des zéros :
        // on conserve la dernière fusion et le prochain appel retentera immédiatement.
        val rtMap = try {
            fetchRealtimeMap()
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            AppLogger.warn("ParcRelaisService temps réel indisponible: ${e.message}")
            return lastMerged ?: staticList
        }
        mutex.withLock { lastRtTimestamp = now }

        val merged = staticList.map { pr ->
            val rawId = pr.id.removePrefix("parc-relais-")
            val dispo = rtMap[rawId]
            if (dispo == null) pr else pr.copy(placesDisponibles = dispo, hasRealtimeData = true,
                etat = if (dispo == 0) ParkingState.COMPLET else ParkingState.OUVERT)
        }
        mutex.withLock { lastMerged = merged }
        return merged
    }

    private suspend fun fetchStatic(): List<Parking> = coroutineScope {
        val resp = safeRequest {
            client.get(staticUrl) {
                timeout { requestTimeoutMillis = NetworkConfiguration.SHARED_TIMEOUT_SECONDS * 1000 }
            }
        }
        if (resp.status != HttpStatusCode.OK) throw ApiError.HttpError(resp.status.value)
        val body: ParcRelaisStaticResponse = safeDecode { resp.body() }
        body.features.mapNotNull { f ->
            val firstPoint = f.geometry.coordinates.firstOrNull() ?: return@mapNotNull null
            if (firstPoint.size < 2) return@mapNotNull null
            val coord = LatLng(firstPoint[1], firstPoint[0])
            Parking(
                id = "parc-relais-${f.properties.id}",
                gid = 0,
                nom = f.properties.nom,
                gestionnaire = "TCL",
                adresse = "",
                latitude = coord.latitude,
                longitude = coord.longitude,
                capaciteTotale = f.properties.capacite ?: 0,
                placesDisponibles = 0,
                etat = ParkingState.OUVERT,
                parkingType = ParkingType.CAR,
                isParcRelais = true,
                horaires = f.properties.horaires,
                surveille = f.properties.pSurv,
                hasRealtimeData = false,
                nbPmr = f.properties.placeHandi
            )
        }
    }

    private suspend fun fetchRealtimeMap(): Map<String, Int> {
        val resp = safeRequest {
            client.get(realtimeUrl) {
                timeout { requestTimeoutMillis = NetworkConfiguration.SHARED_TIMEOUT_SECONDS * 1000 }
            }
        }
        if (resp.status != HttpStatusCode.OK) throw ApiError.HttpError(resp.status.value)
        val body: ParcRelaisRealtimeResponse = safeDecode { resp.body() }
        return body.features.mapNotNull { f ->
            val d = f.properties.nbTotPlaceDispo ?: return@mapNotNull null
            f.properties.id to d
        }.toMap()
    }

    companion object { val shared = ParcRelaisService() }
}
